File: ui-react/src/lib/dsParams.ts
// Parâmetros de execução dos jobs DataStage — lado do front (F3 da spec
// docs/spec-parametros-job-datastage.md).
//
// O que mora aqui: o vocabulário (tipos, origens, âncoras), as réguas de
// validação que ESPELHAM api/services/job_params.py (a API é a fonte da
// verdade — aqui só se evita o 422 e se mostra a mensagem na hora) e a
// conversão entre o DRAFT do editor (tudo string, como o usuário digita) e o
// contrato da API. Sem cálculo de data: a prévia ("com a referência X o valor
// seria Y") vem do servidor (POST /pipelines/jobs/params/preview) — o
// TypeScript NÃO reimplementa o cálculo (decisão da spec §3).
//
// Módulo PURO (sem React) — exercitado pela bancada tests/js/ds_params_harness.cjs.
//
// ⚠️ Regex sempre como LITERAL `/…/`: uma string com escape duplo (`'\\d'`)
// nunca casa e passa tsc + lint + build (gotcha do repo).

export const DS_PARAM_TYPES = [
  'String', 'Integer', 'Float', 'Date', 'Time', 'Timestamp', 'Pathname', 'List', 'Encrypted',
] as const
export type DsParamType = typeof DS_PARAM_TYPES[number]

export const DS_PARAM_SOURCES = [
  { value: 'fixo', label: 'Valor fixo' },
  { value: 'data_referencia', label: 'Data de referência' },
  { value: 'data_logica', label: 'Data lógica (Airflow)' },
  { value: 'data_execucao', label: 'Data da execução' },
  { value: 'run_id', label: 'Run id da corrida' },
] as const
export type DsParamSource = typeof DS_PARAM_SOURCES[number]['value']

export const DS_SOURCES_DATA: readonly string[] = ['data_referencia', 'data_logica', 'data_execucao']
export const DS_TIPOS_COM_DATA: readonly string[] = ['String', 'Date', 'Timestamp']

export const DS_PARAM_ANCORAS = [
  { value: '', label: 'sem âncora' },
  { value: 'inicio_mes', label: 'início do mês' },
  { value: 'fim_mes', label: 'fim do mês' },
  { value: 'inicio_trimestre', label: 'início do trimestre' },
  { value: 'fim_trimestre', label: 'fim do trimestre' },
  { value: 'inicio_ano', label: 'início do ano' },
  { value: 'fim_ano', label: 'fim do ano' },
  { value: 'inicio_semana', label: 'segunda da semana' },
  { value: 'fim_semana', label: 'domingo da semana' },
] as const

export const DS_FORMATOS_SUGERIDOS = [
  '%Y-%m-%d', '%Y%m%d', '%d/%m/%Y', '%Y-%m', '%Y%m', '%Y-%m-%d %H:%M:%S',
]

// Sentinela da API: no GET, o valor Encrypted vem como `***` (+ tem_valor);
// no save, `***` manda preservar o token gravado.
export const DS_ENCRYPTED_MASCARA = '***'
export const DS_LIMITE_NOME = 128
export const DS_LIMITE_FORMATO = 40
export const DS_LIMITE_MESES = 120
export const DS_LIMITE_DIAS = 3660

export const DS_PARAM_NAME_RE = /^[A-Za-z_][A-Za-z0-9_]*(?:\.[A-Za-z_][A-Za-z0-9_]*)?$/
export const DS_FORMATO_RE = /^(?:%[YmdyHMS]|[-/._: ])+$/
const DS_INTEIRO_RE = /^-?\d+$/
const DS_VALOR_FIXO_RE: Record<string, RegExp> = {
  Integer: /^-?\d+$/,
  Float: /^-?\d+(?:\.\d+)?$/,
  Date: /^\d{4}-\d{2}-\d{2}$/,
  Time: /^\d{2}:\d{2}:\d{2}$/,
  Timestamp: /^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$/,
  Pathname: /^\/[^\s'"]+$/,
}

// ── Draft do editor × contrato da API ───────────────────────────────────────

// Linha do editor. storedproc usa só nome/tipo/valor; datastage usa o resto.
// Tudo string (é o que o input entrega); a conversão para a API é aqui.
export interface JobParam {
  id?: string
  param_name: string
  param_type: string
  param_value: string
  param_source?: string
  param_offset_meses?: string
  param_ancora?: string
  param_offset_dias?: string
  param_formato?: string
  // Encrypted com token gravado: valor vazio no draft = "manter".
  tem_valor?: boolean
}

export interface JobParamApi {
  param_name: string
  param_type: string
  param_value: string | null
  param_order?: number
  param_source?: string
  param_offset_meses?: number | null
  param_ancora?: string | null
  param_offset_dias?: number | null
  param_formato?: string | null
  tem_valor?: boolean
}

export function ehOrigemData(origem?: string | null): boolean {
  return DS_SOURCES_DATA.includes(origem ?? '')
}

// GET → draft (vale para storedproc e datastage: os campos extras chegam
// 'fixo'/null no storedproc e não atrapalham). Encrypted: o `***` do GET vira
// campo vazio com tem_valor — o usuário só digita para TROCAR.
export function paramsFromApi(items?: JobParamApi[] | null): JobParam[] {
  return (items ?? []).map((p, i) => {
    const encrypted = p.param_type === 'Encrypted'
    const temValor = encrypted ? !!(p.tem_valor ?? (p.param_value && p.param_value !== '')) : undefined
    return {
      id: `p_${i}_${p.param_name}`,
      param_name: p.param_name,
      param_type: p.param_type,
      param_value: encrypted ? '' : (p.param_value ?? ''),
      param_source: p.param_source ?? 'fixo',
      param_offset_meses: p.param_offset_meses == null ? '' : String(p.param_offset_meses),
      param_ancora: p.param_ancora ?? '',
      param_offset_dias: p.param_offset_dias == null ? '' : String(p.param_offset_dias),
      param_formato: p.param_formato ?? '',
      ...(encrypted ? { tem_valor: temValor } : {}),
    }
  })
}

// storedproc: byte a byte o que a tela sempre mandou (nome/tipo/valor).
export function storedprocParamsToApi(params: JobParam[]): JobParamApi[] {
  return params
    .filter(p => p.param_name.trim())
    .map(p => ({ param_name: p.param_name.trim(), param_type: p.param_type, param_value: p.param_value }))
}

function inteiroOuNulo(v?: string): number | null {
  const s = (v ?? '').trim()
  if (!s) return null
  const n = parseInt(s, 10)
  return Number.isFinite(n) ? n : null
}

// datastage: origem SEMPRE (a API a exige sem default — achado da revisão da
// F1), cálculo só com origem de data (CHECK da migration 107), Encrypted vazio
// com token gravado → `***` (manter).
export function dsParamsToApi(params: JobParam[]): JobParamApi[] {
  return params
    .filter(p => p.param_name.trim())
    .map(p => {
      const origem = p.param_source || 'fixo'
      const data = ehOrigemData(origem)
      let valor: string | null = p.param_value
      if (origem !== 'fixo') valor = null
      else if (p.param_type === 'Encrypted' && !p.param_value && p.tem_valor) valor = DS_ENCRYPTED_MASCARA
      return {
        param_name: p.param_name.trim(),
        param_type: p.param_type,
        param_source: origem,
        param_value: valor,
        param_offset_meses: data ? inteiroOuNulo(p.param_offset_meses) : null,
        param_ancora: data ? ((p.param_ancora ?? '').trim() || null) : null,
        param_offset_dias: data ? inteiroOuNulo(p.param_offset_dias) : null,
        param_formato: data ? ((p.param_formato ?? '').trim() || null) : null,
      }
    })
}

// Despacho por tipo de etapa — o que vai na chave `params` do save.
export function paramsToApi(jobType: string, params: JobParam[]): JobParamApi[] {
  if (jobType === 'storedproc') return storedprocParamsToApi(params)
  if (jobType === 'datastage') return dsParamsToApi(params)
  return []
}

// Itens para a PRÉVIA: os mesmos da API, mas o segredo NUNCA sai para a prévia
// (ela devolve `***` de qualquer forma) — manda um marcador no lugar.
export function previewItens(params: JobParam[]): JobParamApi[] {
  return dsParamsToApi(params).map(p => (
    p.param_type === 'Encrypted' && p.param_source === 'fixo'
      ? { ...p, param_value: p.param_value ? 'x' : '' }
      : p
  ))
}

// ── Validação (espelho de normalizar_item/normalizar_lista da API) ──────────

function inteiroNaFaixa(v: string | undefined, rotulo: string, limite: number, erros: string[]) {
  const s = (v ?? '').trim()
  if (!s) return
  if (!DS_INTEIRO_RE.test(s)) { erros.push(`${rotulo} inválido — informe um inteiro`); return }
  if (Math.abs(parseInt(s, 10)) > limite) erros.push(`${rotulo} fora da faixa ±${limite}`)
}

export function dsParamErrors(params: JobParam[]): string[] {
  const erros: string[] = []
  const vistos = new Set<string>()
  params.forEach((p, i) => {
    const nome = p.param_name.trim()
    const rotulo = nome ? `Parâmetro "${nome}"` : `Parâmetro #${i + 1}`
    // Linha totalmente vazia = linha do editor ainda não preenchida (a API a ignora).
    if (!nome && !(p.param_value ?? '').trim()) return
    if (!nome || !DS_PARAM_NAME_RE.test(nome)) {
      erros.push(`${rotulo}: nome inválido — letras/números/_ e no máximo um ponto (PSet.Param)`)
    } else if (nome.length > DS_LIMITE_NOME) {
      erros.push(`${rotulo}: nome com ${nome.length} caracteres — máximo ${DS_LIMITE_NOME}`)
    }
    if (!(DS_PARAM_TYPES as readonly string[]).includes(p.param_type)) {
      erros.push(`${rotulo}: tipo inválido`)
    }
    const origem = p.param_source || ''
    if (!origem || !DS_PARAM_SOURCES.some(s => s.value === origem)) {
      erros.push(`${rotulo}: escolha a origem do valor`)
      return
    }
    if (ehOrigemData(origem)) {
      if (!DS_TIPOS_COM_DATA.includes(p.param_type)) {
        erros.push(`${rotulo}: origem de data só com tipo String, Date ou Timestamp`)
      }
      inteiroNaFaixa(p.param_offset_meses, `${rotulo}: meses`, DS_LIMITE_MESES, erros)
      inteiroNaFaixa(p.param_offset_dias, `${rotulo}: dias`, DS_LIMITE_DIAS, erros)
      const ancora = (p.param_ancora ?? '').trim()
      if (ancora && !DS_PARAM_ANCORAS.some(a => a.value === ancora)) erros.push(`${rotulo}: âncora inválida`)
      const formato = (p.param_formato ?? '').trim()
      if (formato && !DS_FORMATO_RE.test(formato)) {
        erros.push(`${rotulo}: formato inválido — só %Y %m %d %y %H %M %S e separadores - / . _ :`)
      } else if (formato.length > DS_LIMITE_FORMATO) {
        erros.push(`${rotulo}: formato com ${formato.length} caracteres — máximo ${DS_LIMITE_FORMATO}`)
      }
    } else if (origem === 'run_id') {
      if (p.param_type !== 'String') erros.push(`${rotulo}: origem run_id só com tipo String`)
    } else {
      const valor = p.param_value ?? ''
      if (p.param_type === 'Encrypted') {
        if (!valor && !p.tem_valor) erros.push(`${rotulo}: informe o valor (Encrypted)`)
      } else if (!valor && p.param_type !== 'String') {
        erros.push(`${rotulo}: valor obrigatório para origem fixa`)
      } else if (/[\r\n]/.test(valor)) {
        erros.push(`${rotulo}: valor não pode ter quebra de linha`)
      } else if (DS_VALOR_FIXO_RE[p.param_type] && !DS_VALOR_FIXO_RE[p.param_type].test(valor)) {
        erros.push(`${rotulo}: "${valor}" não é um ${p.param_type} válido`)
      }
    }
    // Duplicata por CAIXA EXATA: o DataStage distingue pData de pdata.
    if (nome) {
      if (vistos.has(nome)) erros.push(`${rotulo}: duplicado`)
      vistos.add(nome)
    }
  })
  return erros
}

File: dags/utils/ds_params.py
"""
dags/utils/ds_params.py — parâmetros de execução dos jobs DataStage: o que o
DataStageOperator manda em `-param` (spec docs/spec-parametros-job-datastage.md, F2).

⚠️ ESPELHO de api/services/job_params.py. O vocabulário e as funções de cálculo
(deslocar_meses, ancorar, calcular_data, formatar, descrever, parse_data) são
cópia byte a byte: api/ e dags/ rodam em containers diferentes e o repo nunca
importou uma árvore da outra. tests/test_job_params_antidrift.py cobra valor E
descrição iguais nas duas. Mudou lá, muda aqui.

O que só existe aqui (runtime):

  carregar_etapa   linhas da etapa em etl_pipeline_job_param (placeholders %s — a
                   árvore dags/ fala pymssql; api/ fala pyodbc com `?`)
  parse_lparams    nomes que o job DECLARA (saída do `dsjob -lparams`)
  mesclar          pipeline (só o que o job declara) < etapa (não declarado = ERRO)
                   < sobreposição do rerun — o contrato do §4 da spec
  resolver         origem → valor (Encrypted decifrado com a ORQUESTRA_CONN_KEY),
                   com fonte e descrição
  montar_args      os pares `-param nome=valor`, quotados para o shell remoto
  linha_de_log / para_json   o RASTRO do que foi enviado — Encrypted SEMPRE `***`

Tudo puro exceto carregar_etapa (recebe o hook). A validação de formato/tipo NÃO
mora aqui: ela é da API (e dos CHECKs da migration 107) — o runtime confia no que
o banco guardou e falha alto no que não consegue resolver.
"""
from __future__ import annotations

import calendar
import json
import os
import re
import shlex
from datetime import date, datetime, timedelta

# ═══════════════════════════════════════════════════════════════════════════
# Vocabulário e cálculo — ESPELHO de api/services/job_params.py
# ═══════════════════════════════════════════════════════════════════════════

DS_PARAM_TYPES = ("String", "Integer", "Float", "Date", "Time", "Timestamp",
                  "Pathname", "List", "Encrypted")
DS_PARAM_SOURCES = ("fixo", "data_referencia", "data_logica", "data_execucao", "run_id")
DS_SOURCES_DATA = ("data_referencia", "data_logica", "data_execucao")
DS_PARAM_ANCORAS = ("inicio_mes", "fim_mes", "inicio_trimestre", "fim_trimestre",
                    "inicio_ano", "fim_ano", "inicio_semana", "fim_semana")
FORMATO_PADRAO = "%Y-%m-%d"
ENCRYPTED_MASCARA = "***"

ROTULO_ORIGEM = {
    "data_referencia": "referência",
    "data_logica": "data lógica",
    "data_execucao": "data da execução",
}
ROTULO_ANCORA = {
    "inicio_mes": "início do mês",
    "fim_mes": "fim do mês",
    "inicio_trimestre": "início do trimestre",
    "fim_trimestre": "fim do trimestre",
    "inicio_ano": "início do ano",
    "fim_ano": "fim do ano",
    "inicio_semana": "segunda da semana",
    "fim_semana": "domingo da semana",
}


def deslocar_meses(d: date, meses: int) -> date:
    """`d` deslocada `meses` meses, com o dia TRUNCADO ao último válido do mês
    de destino: 31/03 −1 = 28/02 (29/02 em bissexto); 30/04 +1 = 30/05."""
    if not meses:
        return d
    total = d.year * 12 + (d.month - 1) + meses
    ano, mes0 = divmod(total, 12)
    mes = mes0 + 1
    ultimo = calendar.monthrange(ano, mes)[1]
    return date(ano, mes, min(d.day, ultimo))


def ancorar(d: date, ancora: str | None) -> date:
    """Leva `d` ao ponto da âncora dentro do próprio período. Semana = seg–dom."""
    if not ancora:
        return d
    if ancora == "inicio_mes":
        return d.replace(day=1)
    if ancora == "fim_mes":
        return d.replace(day=calendar.monthrange(d.year, d.month)[1])
    if ancora == "inicio_trimestre":
        return date(d.year, 3 * ((d.month - 1) // 3) + 1, 1)
    if ancora == "fim_trimestre":
        mes = 3 * ((d.month - 1) // 3) + 3
        return date(d.year, mes, calendar.monthrange(d.year, mes)[1])
    if ancora == "inicio_ano":
        return date(d.year, 1, 1)
    if ancora == "fim_ano":
        return date(d.year, 12, 31)
    if ancora == "inicio_semana":
        return d - timedelta(days=d.weekday())
    if ancora == "fim_semana":
        return d + timedelta(days=6 - d.weekday())
    raise ValueError(f"âncora desconhecida: {ancora!r}")


def calcular_data(base: date, meses: int = 0, ancora: str | None = None, dias: int = 0) -> date:
    """O contrato do §4: meses → âncora → dias. Cada passo é opcional."""
    d = deslocar_meses(base, int(meses or 0))
    d = ancorar(d, ancora or None)
    if dias:
        d = d + timedelta(days=int(dias))
    return d


def formatar(d: date, formato: str | None) -> str:
    return d.strftime(formato or FORMATO_PADRAO)


def _plural(n: int, singular: str, plural: str) -> str:
    n = int(n)
    sinal = "+" if n > 0 else "-"
    return f"{sinal}{abs(n)} {singular if abs(n) == 1 else plural}"


def descrever(base: date, origem: str, meses: int = 0, ancora: str | None = None,
              dias: int = 0, formato: str | None = None) -> str:
    """Texto humano do cálculo, o MESMO que vai para o log da task, para a
    prévia da tela e para o modal de rerun — é o rastro de "qual parâmetro
    foi usado". Ex.: `referência 2026-09-09 → -1 mês → fim do mês → 2026-08-31`."""
    partes = [f"{ROTULO_ORIGEM.get(origem, origem)} {base.isoformat()}"]
    if meses:
        partes.append(_plural(meses, "mês", "meses"))
    if ancora:
        partes.append(ROTULO_ANCORA.get(ancora, ancora))
    if dias:
        partes.append(_plural(dias, "dia", "dias"))
    partes.append(formatar(calcular_data(base, meses, ancora, dias), formato))
    return " → ".join(partes)


def parse_data(valor) -> date | None:
    """'YYYY-MM-DD' (ou date/datetime) → date; None quando inválido."""
    if isinstance(valor, datetime):
        return valor.date()
    if isinstance(valor, date):
        return valor
    try:
        return datetime.strptime(str(valor or "").strip(), "%Y-%m-%d").date()
    except ValueError:
        return None


# ═══════════════════════════════════════════════════════════════════════════
# Runtime — só existe aqui
# ═══════════════════════════════════════════════════════════════════════════

class ParamError(Exception):
    """Motivo pelo qual a etapa deve FALHAR antes do disparo (o operador embrulha
    em AirflowException com o prefixo [DS]). Nunca dispara "sem o parâmetro"."""


ENV_KEY = "ORQUESTRA_CONN_KEY"

# Colunas na ORDEM do SELECT — o dict de cada linha é o que mesclar/resolver leem.
COLS_ETAPA = ("param_name", "param_type", "param_value", "param_source",
              "param_offset_meses", "param_ancora", "param_offset_dias", "param_formato")
SQL_ETAPA = (
    "SELECT param_name, param_type, param_value, param_source, param_offset_meses, "
    "param_ancora, param_offset_dias, param_formato "
    "FROM dbo.etl_pipeline_job_param WHERE pipeline_name=%s AND job_name=%s "
    "ORDER BY param_order")
_SEM_107_RE = re.compile(r"invalid column name|invalid object name", re.IGNORECASE)
# F4 — defaults do pipeline (migration 108): mesmas colunas, mesma ordem.
SQL_PIPELINE = (
    "SELECT param_name, param_type, param_value, param_source, param_offset_meses, "
    "param_ancora, param_offset_dias, param_formato "
    "FROM dbo.etl_pipeline_param WHERE pipeline_name=%s ORDER BY param_order")


def _decrypt(token: str) -> str:
    """Token Fernet → valor. Mesma chave e mesma disciplina de
    dags/utils/conn_resolver._decrypt (que não é importado daqui para não
    arrastar os hooks do Airflow no import). O retorno NUNCA vai para log."""
    from cryptography.fernet import Fernet  # dependência do próprio Airflow
    key = (os.getenv(ENV_KEY) or "").strip()
    if not key:
        raise ParamError(
            f"{ENV_KEY} não configurada no worker — defina no .env/compose o MESMO "
            "valor usado pelo orquestra-api (x-airflow-common); sem ela um parâmetro "
            "Encrypted não pode ser decifrado")
    try:
        return Fernet(key.encode()).decrypt(str(token).encode("ascii")).decode("utf-8")
    except Exception as e:  # InvalidToken, base64 ruim…
        raise ParamError(
            "valor Encrypted ilegível — a ORQUESTRA_CONN_KEY do worker não corresponde "
            "à chave com que o parâmetro foi salvo no orquestra-api; redefina o valor "
            "na etapa ou restaure a chave") from e


def carregar_etapa(hook, pipeline_name: str, job_name: str, log=None) -> list[dict]:
    """Linhas de etl_pipeline_job_param da etapa, na ordem do editor.

    Sem a migration 107 (coluna param_source ausente) não pode haver parâmetro
    de etapa datastage — a API recusa gravar — então a resposta honesta é
    "nenhum", com aviso. Qualquer OUTRO erro de banco propaga: sem saber os
    parâmetros, disparar "sem eles" seria o falso verde que a spec proíbe."""
    try:
        rows = hook.get_records(SQL_ETAPA, parameters=(pipeline_name, job_name))
    except Exception as e:
        if _SEM_107_RE.search(str(e)):
            if log is not None:
                log.warning("[DS] etl_pipeline_job_param sem as colunas da migration 107 "
                            "(%s) — nenhum parâmetro de etapa será enviado", e)
            return []
        raise
    return [dict(zip(COLS_ETAPA, r)) for r in (rows or [])]


def carregar_pipeline(hook, pipeline_name: str, log=None) -> list[dict]:
    """Defaults do pipeline (etl_pipeline_param, F4), na ordem do editor.

    Sem a migration 108 (tabela ausente) não há default nenhum — resposta
    honesta é "nenhum", em debug (é o estado normal até a F4 ir para
    produção). Qualquer OUTRO erro de banco propaga, como em carregar_etapa."""
    try:
        rows = hook.get_records(SQL_PIPELINE, parameters=(pipeline_name,))
    except Exception as e:
        if _SEM_107_RE.search(str(e)):
            if log is not None:
                log.debug("[DS] etl_pipeline_param ausente (migration 108) — sem defaults do pipeline (%s)", e)
            return []
        raise
    return [dict(zip(COLS_ETAPA, r)) for r in (rows or [])]


def parse_lparams(saida: str) -> set[str]:
    """Nomes que o job declara, um por linha no `dsjob -lparams` — parâmetros
    simples e membros de Parameter Set (`PSet.Param`). Ignora a linha
    `Status code = N` (o mesmo parser do Console DataStage)."""
    nomes: set[str] = set()
    for raw in (saida or "").splitlines():
        linha = raw.strip()
        if not linha or re.match(r"^status code\s*=", linha, re.IGNORECASE):
            continue
        nomes.add(linha)
    return nomes


def mesclar(etapa: list[dict], declarados: set[str], pipeline: list[dict] | None = None,
            overrides: list[dict] | None = None) -> tuple[list[dict], list[str]]:
    """O contrato do §4 (passos 1–3): devolve (itens, ignorados_do_pipeline).

    • pipeline: só o que o job DECLARA entra; o resto vai para `ignorados`
      (e para o log) — default compartilhado por natureza.
    • etapa: sobrepõe por nome; nome NÃO declarado = ParamError listando o que
      o job declara (o operador digitou o nome de propósito).
    • overrides (F5, {param_name, param_value}): sobrepõem o valor como
      `fixo`; nome fora de tudo e não declarado = ParamError.
    Cada item sai com `fonte` ('pipeline' | 'etapa' | 'rerun')."""
    por_nome: dict[str, dict] = {}
    ignorados: list[str] = []
    for p in pipeline or []:
        if p["param_name"] in declarados:
            por_nome[p["param_name"]] = dict(p, fonte="pipeline")
        else:
            ignorados.append(p["param_name"])
    nao_declarados = [p["param_name"] for p in etapa if p["param_name"] not in declarados]
    if nao_declarados:
        raise ParamError(
            "o job NÃO declara o(s) parâmetro(s) cadastrado(s) na etapa: "
            + ", ".join(nao_declarados)
            + ". O DataStage distingue maiúsculas de minúsculas. Parâmetros que o job declara: "
            + (", ".join(sorted(declarados)) if declarados else "(nenhum)"))
    for p in etapa:
        por_nome[p["param_name"]] = dict(p, fonte="etapa")
    for o in overrides or []:
        nome = o["param_name"]
        if nome not in por_nome and nome not in declarados:
            raise ParamError(
                f"a sobreposição do rerun cita '{nome}', que o job não declara. "
                "Parâmetros que o job declara: " + (", ".join(sorted(declarados)) or "(nenhum)"))
        base = por_nome.get(nome, {"param_name": nome, "param_type": "String"})
        por_nome[nome] = dict(base, param_source="fixo", param_value=o.get("param_value"),
                              param_offset_meses=None, param_ancora=None,
                              param_offset_dias=None, param_formato=None, fonte="rerun")
    return list(por_nome.values()), ignorados


def resolver(itens: list[dict], bases: dict, run_id: str, decifrar=None) -> list[dict]:
    """Passo 4–5 do §4: cada item → {name, valor, fonte, descricao, mascarado}.

    `bases` = {origem_de_data: date} já resolvidas pelo operador (ele só busca
    a data de referência no banco quando algum item precisa). Encrypted é
    decifrado aqui e marcado `mascarado=True` — quem loga/persiste usa
    valor_exibido()."""
    decifrar = decifrar or _decrypt
    saida: list[dict] = []
    for it in itens:
        nome = it["param_name"]
        tipo = it.get("param_type") or "String"
        origem = it.get("param_source") or "fixo"
        mascarado = False
        if origem in DS_SOURCES_DATA:
            base = bases.get(origem)
            if base is None:
                raise ParamError(f"'{nome}': origem {origem} indisponível nesta execução")
            meses = it.get("param_offset_meses") or 0
            dias = it.get("param_offset_dias") or 0
            ancora = it.get("param_ancora") or None
            formato = it.get("param_formato") or None
            try:
                valor = formatar(calcular_data(base, meses, ancora, dias), formato)
                descricao = descrever(base, origem, meses, ancora, dias, formato)
            except (ValueError, OverflowError) as e:
                raise ParamError(f"'{nome}': data fora do calendário ({e})") from e
        elif origem == "run_id":
            valor, descricao = str(run_id or ""), "run_id da corrida"
        else:  # fixo
            valor = it.get("param_value")
            valor = "" if valor is None else str(valor)
            if tipo == "Encrypted":
                if not valor:
                    raise ParamError(f"'{nome}': Encrypted sem valor gravado")
                valor = decifrar(valor)
                descricao, mascarado = "fixo (Encrypted)", True
            else:
                descricao = "fixo"
        saida.append({"name": nome, "valor": valor, "fonte": it.get("fonte") or "etapa",
                      "descricao": descricao, "mascarado": mascarado})
    return saida


def valor_exibido(p: dict) -> str:
    return ENCRYPTED_MASCARA if p.get("mascarado") else str(p.get("valor", ""))


def montar_args(lista: list[dict], exibir: bool = False) -> list[str]:
    """`-param nome=valor` por item, quotado pelo shlex para o shell remoto
    (só ganha aspas quando precisa). `exibir=True` monta a versão para log e
    mensagens de erro — Encrypted vira `***`."""
    args: list[str] = []
    for p in lista:
        v = valor_exibido(p) if exibir else str(p.get("valor", ""))
        args += ["-param", shlex.quote(f"{p['name']}={v}")]
    return args


def linha_de_log(lista: list[dict], ignorados: list[str] | None = None) -> str:
    """A linha `[DS] parâmetros:` — nome=valor (fonte · descrição), Encrypted
    mascarado, mais os defaults de pipeline que o job não declara."""
    partes = [f"{p['name']}={valor_exibido(p)} ({p.get('fonte', 'etapa')} · {p.get('descricao', '')})"
              for p in lista]
    texto = " · ".join(partes) if partes else "nenhum"
    if ignorados:
        texto += " · ignorados do pipeline (o job não declara): " + ", ".join(ignorados)
    return texto


def para_json(lista: list[dict]) -> str:
    """etl_ds_job_log.params_json — [{name, valor, fonte, descricao, mascarado}],
    com o valor JÁ mascarado quando Encrypted (o segredo nunca vai ao banco em claro)."""
    return json.dumps([{"name": p["name"], "valor": valor_exibido(p),
                        "fonte": p.get("fonte", "etapa"), "descricao": p.get("descricao", ""),
                        "mascarado": bool(p.get("mascarado"))} for p in lista],
                      ensure_ascii=False)
